src/v2.rs

```diff
@@ -0,0 +1,676 @@
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::future::Future;
+use std::mem::size_of;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::task::Poll;
+use std::task::Waker;
+
+use deno_ast::EmitOptions;
+use deno_ast::TranspiledSource;
+use deno_graph::ModuleGraph;
+use futures::future::poll_fn;
+use sha2::Digest;
+use sha2::Sha256;
+use tokio::io::AsyncReadExt;
+
+use crate::error::ParseError;
+use crate::Module;
+use crate::ModuleInner;
+use crate::ModuleKind;
+
+pub(crate) const ESZIP_V2_MAGIC: &[u8; 8] = b"ESZIP_V2";
+
+#[derive(Debug, PartialEq)]
+#[repr(u8)]
+enum HeaderFrameKind {
+  Module = 0,
+  Redirect = 1,
+}
+
+#[derive(Debug, Default)]
+pub struct EsZipV2 {
+  modules: Arc<Mutex<HashMap<String, EszipV2Module>>>,
+}
+
+#[derive(Debug)]
+enum EszipV2Module {
+  Module {
+    kind: ModuleKind,
+    source: EszipV2SourceSlot,
+    source_map: EszipV2SourceSlot,
+  },
+  Redirect {
+    target: String,
+  },
+}
+
+#[derive(Debug)]
+enum EszipV2SourceSlot {
+  Pending {
+    offset: usize,
+    length: usize,
+    wakers: Vec<Waker>,
+  },
+  Ready(Arc<Vec<u8>>),
+}
+
+impl EszipV2SourceSlot {
+  fn bytes(&self) -> &[u8] {
+    match self {
+      EszipV2SourceSlot::Ready(v) => v,
+      _ => panic!("EszipV2SourceSlot::unwrap() called on a pending slot"),
+    }
+  }
+}
+
+impl EsZipV2 {
+  pub async fn parse<R: tokio::io::AsyncRead + Unpin>(
```

Comment:
IMO I would prefer to keep the header logic in the tokio_util::codec from #44 
`EsZIpV2::parse` could just collective parse the header and use the logic for the future.

Reply:
@littledivy Now that the header does not need to be streamed anymore, `tokio_util::codec` is a bit overkill. We never have the "failure condition" where we run out of bytes, because all header bytes are available upfront. Because of that, it is confusing that the decoding can return `None` for both "out of bytes" which should never happen anymore unless the header is malformed, and "we're done here" which we don't need because we already know the exact length of the header upfront.

I think the current code is reasonable enough to keep.

Reply:
:+1: Makes sense. disregard my comment.